### DISCUSSION TITLE
Jinja variables in reference links

### DISCUSSION THREAD

**Discussion by redtide** [mkdocs/mkdocs#3185] [Q&A]
I haven't found how to use variables on reference links (not even inline). Is it possible?

E.g.:

```markdown
read [this page] how to...
...

[this page]: {{ base_url }}/somedir/somepage
```

**Comment by kamilkrzyskow** [comment]
{{ base_url }} looks like Jinja2 syntax, and the rest of the content seems to be Markdown.
To be able to use Jinja2 variables in Markdown you'd have to use the macros plugin:
https://mkdocs-macros-plugin.readthedocs.io/en/latest/

**Reply by redtide** [reply]
I knew about it but I wasn't able to make it work for something wrong I did, glad I can use it like with jekyll.

**Reply by redtide** [reply]
I spoke too soon, macros plugin doesn't seems to help to retrieve variables made [by mkdocs](https://github.com/mkdocs/mkdocs/blob/master/mkdocs/commands/build.py#L49), or is not clear how to use it in this case, see also [url()](https://www.mkdocs.org/dev-guide/themes/#url).

**Reply by kamilkrzyskow** [reply]
Oh wow, first time getting an answer unmarked 👀💦 
Looking at the docs/code, seems you're right and there is no direct access to the real context variables like `base_url`, because they aren't even created yet. I never worked with the plugin myself, but it always seemed like it allows to do almost everything. 

I guess you'd need to create a [module](https://mkdocs-macros-plugin.readthedocs.io/en/latest/macros/), which would expose a recreated copy of either the `base_url` or `url` filter function. 

The difficulty has risen quite substantially and probably won't be as "handy", as you want. Perhaps this is a https://xyproblem.info/
and your goal isn't exactly getting access to these variables, but maybe you're just lazy to type out relative paths in Markdown 🤔
I recall there was a plugin for that which automatically creates relatives paths to any resource, as long the name is unique, can't recall the name currently.

Perhaps I missed the mark and you want access to those variables to do something more elaborate than path creation.

**Reply by redtide** [reply]
> Oh wow, first time getting an answer unmarked

I'm sorry for that, I hoped that was simple like write a python macro but I forgot that was about an inaccessible information that mkdocs have

> I guess you'd need to create a [module](https://mkdocs-macros-plugin.readthedocs.io/en/latest/macros/), which would expose a recreated copy of either the `base_url` or `url` filter function.

that was an idea but IDK how, I'm new to this and in the end it seems just a workaround, would be nice if it was a mkdocs feature (unless I had a solution for an accepted PR)

> Perhaps this is a https://xyproblem.info/ and your goal isn't exactly getting access to these variables, but maybe you're just lazy to type out relative paths in Markdown

TLDR: I need to know the real root path of the site, so please don't be arrogant, I'm not a lazy ass.

or maybe I  I'm tired to have wrong links when forking `someproject.github.io` resulting in a `redtide.github.io/someproject.github.io/` (or else), so many relative links don't work. In jekyll other than using `baseurl` I had to use things like `{{ '/root/relative/link` | relative_url }}`.

> I recall there was a plugin for that which automatically creates relatives paths to any resource, as long the name is unique, can't recall the name currently.

heh, what unfortunate, but maybe I found what I need with `config.site_url`.

**Reply by kamilkrzyskow** [reply]
Oh don't worry about the unmark, if it's not the answer it rightfully shouldn't be marked so it's 👌 
Don't take such lightly thrown words to heart, but if you feel insulted, I'm sorry for the low joke ✌️ 

Whenever I fork projects they have the same structure on my repository, and hosting GitHub Pages works the same as on the original repository, so any relative link works 🤔 
I guess it's an issue with the "domain" projects `username.github.io` as it assumes a `/` root path, but as you have your own `redtide.github.io` it creates a subdirectory fork, which does break the paths 😒
Then again I'm not sure if that's the case with relative paths, `../file.md`, it doesn't matter if the projects is hosted on `/` or `/project_name/` URL, the distance to the root directory should always be the same... I feel like I'm still missing something about the problem.

You plan to access `config.site_url` via the `macros` plugin right? But won't that require setting this method up in every fork with broken relative paths? If that's correct then it's quite the ordeal, because it's not a one time solution, but rather requires maintenance on each new fork.

I don't have better ideas ATM, so good luck

**Reply by redtide** [reply]
> Then again I'm not sure if that's the case with relative paths, `../file.md`, it doesn't matter if the projects is hosted on `/` or `/project_name/` URL, the distance to the root directory should always be the same... I feel like I'm still missing something about the problem.

Maybe I'm the one that misses something, I don't recall the reason I started this instead simple relative paths from the current file, I had to document the case but I didn't.

> You plan to access `config.site_url` via the `macros` plugin right? But won't that require setting this method up in every fork with broken relative paths? If that's correct then it's quite the ordeal, because it's not a one time solution, but rather requires maintenance on each new fork.

If it was the solution I had in mind, the one I used with Jekyll, it wouldn't require any maintenance, but unfortunately it isn't, it can just make either a local `http://127.0.0.1:8000` or the link hardcoded in config, not the real one.

> I don't have better ideas ATM, so good luck

Thank you for help and your time
